### PR TITLE
Stencil reads are S,X,X,X. (but expect S,S,S,S or S,0,0,1)

### LINKED ISSUE #1198: 2nd, 3rd, and 4th components of sampled depth/stencil textures are undefined in Metal
In WGSL, performing a `textureSample` or `textureLoad` operation on a `texture_sampled_2d<type>` yields a `vec4<type>`. This is true even for textures that have fewer than 4 components ex.) `r8uint`, `r16float`, `rg8unorm`. It is possible to bind a texture with `n` components to `texture_sampled_2d<type>` as long as the texture's component type matches the `type` in the shader. You can bind a texture of type `depth32float` to `texture_sampled_2d<float>`.

Sampling the "unspecified" components yields default values. i.e. accessing the `b` component of an `rg` texture yields some default value.

According to the Metal Shading Language Specification (EDIT:) 2.3 section 6.10 / 2.4 section 6.12:
> For unspecified color components in a pixel format, the default values are:
> • zero, for components other than alpha.
> • one, for the alpha component.
> ...
> ...
> **For a texture with depth or stencil pixel format
> (such as MTLPixelFormatDepth24Unorm_Stencil8 or MTLPixelFormatStencil8), the
> default value for an unspecified component is undefined**

On the contrary, in all of the other APIs (Vulkan, D3D12, OpenGL too), when sampling a depth texture (it has 1 component), accessing the "unspecified" components `.g`, `.b`, and `.a` yield 0, 0, and 1, respectively.

In recent versions of MacOS (10.15+) and iOS (13.0+), `MTLTextureSwizzleChannels` can be set on texture view creation to force the unspecified components to some component index, or 0, or 1. Unfortunately, on older Apple platforms, mitigating the undefined values in Metal would involve binding extra data to the shader to indicate whether at draw-time a texture is a depth/stencil texture in order to emulate correct values for the unspecified components. We could also recompile pipelines and pass the data as MTLFunctionConstants, but that may be a worse solution.

If we want to avoid binding extra data per-sampled-texture or recompiling pipelines, an alternative solution could be to add _even more_ binding types and WGSL types: `sampled-depth-texture` and `sampled-stencil-texture`. This way, a WGSL compiler knows at compile-time if a texture is a depth/stencil texture and can insert shader code to emulate the unspecified components.

_Note: When I say "sampling a depth texture" I mean sampling the values out of a depth texture bound to `texture_sampled_2d<f32>`. This is not the same as using `texture_depth_2d` which is for comparison sampling._

### PR DESCRIPTION
**Pull request by jdashg**
Closes #1198.

***
<a href="https://pr-preview.s3.amazonaws.com/jdashg/gpuweb/pull/1978.html" title="Last updated on Aug 4, 2021, 5:11 PM UTC (7307ee0)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/1978/7439a4d...jdashg:7307ee0.html" title="Last updated on Aug 4, 2021, 5:11 PM UTC (7307ee0)">Diff</a>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -9532,38 +9533,51 @@ None of the depth formats can be filtered.
         <td>1 &minus; 4
         <td>stencil
         <td>{{GPUTextureSampleType/"uint"}}
+        <td>`vec4<u32>(S, S, S, S)` or `vec4<u32>(S, 0, 0, 1)`
```
Comment (kvark) [inline spec/index.bs:9536]: can also write `vec4<u32>(S)`

File: spec/index.bs [hunk 2]
```diff
@@ -9524,6 +9524,7 @@ None of the depth formats can be filtered.
             <th>Bytes per texel
             <th>Aspect
             <th>{{GPUTextureSampleType}}
+            <th>Returned as...
```
Comment (kvark) [inline spec/index.bs:9527]: let's clarify that the "Returned" refers to the shader specifically

**Review by kvark — COMMENTED** [review]
Looks fine, assuming that we can indeed rely on these two possibilities.

**Comment by kainino0x** [comment]
Resolution: Make this non-normative because Metal doesn't actually guarantee this.

**Comment by jdashg** [comment]
> Resolution: Make this non-normative because Metal doesn't actually guarantee this.

More specifically, our consensus is that it's not worth it to try to make this normative. (not just because Metal doesn't guarantee it per se) Making it normative can exert pressure to keep just these two possibilities, so that no one ends up adding a third. (like S,0,0,0?) Probably not a big concern.
Myles additionally noted that making this normative probably doesn't help authors, and that authors are our primary audience.

**Review by kainino0x — APPROVED** [review]
LGTM. I think we also needed to do something like this for binding depth textures as non-depth textures? But we can follow up with it.

**Comment by kvark** [comment]
> LGTM. I think we also needed to do something like this for binding depth textures as non-depth textures? But we can follow up with it.

I thought the response from Apple was that it's just UB and we aren't supposed to do this at all. (i.e. it's not just the swizzling problem).

**Comment by kainino0x** [comment]
Oh yeah, I think you're right. Can't keep all this stuff in my head

File: spec/index.bs [hunk 3]
```diff
@@ -9532,38 +9533,52 @@ None of the depth formats can be filtered.
...
         <td rowspan=2 style='white-space:nowrap'>{{GPUTextureFormat/depth24plus-stencil8}}
         <td rowspan=2>4 &minus; 8
         <td>depth
         <td>{{GPUTextureSampleType/"depth"}}
+        <td>`f32(D)`
         <td colspan=2>&cross;
     <tr>
         <td>stencil
         <td>{{GPUTextureSampleType/"uint"}}
+        <td>`vec4<u32>(S, S, S, S)` or `vec4<u32>(S, 0, 0, 1)`
```
Comment (kainino0x) [inline spec/index.bs:9562]: Myles points out this still needs to be updated.

File: spec/index.bs (hunk 1 above)
Comment (jdashg) [reply spec/index.bs:9536]: This needs to be S,X,X,X

Comment (litherum) [reply spec/index.bs:9536]: I don't think this matches the prose below.

File: spec/index.bs (hunk 3 above)
Comment (litherum) [inline spec/index.bs:9562]: ditto.

**Pull request merged**